use mailing_list_emails view instead of looping through all couch users

proteusvacuum opened the pull request:
http://manage.dimagi.com/default.asp?167775#935442
@gcapalbo 
FYI: @snopoke

proteusvacuum committed:
use mailing_list_emails view instead of looping through all couch users

snopoke commented on corehq/apps/registration/utils.py:
```diff
@@ -98,7 +98,11 @@ def is_user_subscribed_with_email(couch_user):
         other_user
-        for other_user in CouchUser.all()
+        for other_user in CouchUser.view(
+            'users/mailing_list_emails',
```
You should add `key=old_email` to restrict the results to only those users with `email = old_email`.

proteusvacuum committed:
use couch to filter results

snopoke commented:
Looks good. Were you able to test this at all?

gcapalbo commented:
:+1:

proteusvacuum committed:
handling email changes with mailchimp lists

- Use postgres users table to find other users with the same email
- Previously, the `handle_changed_mailchimp_email` method didn't properly unsubscribe people from the MASS_EMAIL mailing list
- Also, adds some simple tests to make sure that if there is more than one person with the same email address in the system, the email is not removed from the mailinglist

proteusvacuum committed:
checks whether other user is subscribed to mailchimp

proteusvacuum committed:
pull all values from couchdb at the same time

proteusvacuum commented:
@snopoke, I re-wrote this. I think this makes a lot more sense now. Sorry for the number of commits on the open PR, messing around with couch.

proteusvacuum commented on corehq/apps/registration/utils.py:
```diff
@@ -92,27 +93,35 @@ def safe_unsubscribe_user_from_mailchimp_list(user, list_id, email=None):
+                                          reduce=False,
+                                          ).all()
+    for couch_user in couch_users:
+        if couch_user['doc']['subscribed_to_commcare_users'] or not couch_user['doc']['email_opt_out']:
```
Couch question: can I be sure that a `CouchUser` will have these keys? Is it necessary to be safe with `couch_user.get('doc', {}).get('subscribed_to_commcare_users')`?

proteusvacuum replied:
Also, would the Pythonic way here be to use a list comprehension.?

```py
def subscribed(user):
    return user.get('doc',{}).get('subscribed_to_commcare_users') or not user.get('doc',{}).get('email_opt_out', True)
users_subscribed_with_old_email = [couch_user.get('id') for couch_user in couch_users if subscribed(couch_user)]
```

czue replied:
for the first question, you can't necessarily make that assumption if you're working with the raw JSON coming out of the database. If you work with the wrapped objects (so `CouchUser.view` instead of `CouchUser.get_db().view` which will automatically unpack the doc from the result and wrap it with the class) then `jsonobject` will handle the wrapping for you and let you reference `user.subscribed_to_commcare_users`). That said it's a performance tradeoff since wrapping takes additional work. In this case since we expect the number of docs to be quite small, wrapping is probably fine. 

Also I like list comprehensions, but I don't think there's an explicit style guide around that.

snopoke replied:
+1

czue commented on corehq/apps/registration/tests/utils.py:
```diff
@@ -0,0 +1,75 @@
+from corehq.apps.registration.utils import handle_changed_mailchimp_email
+
+
+class UtilTestCase(TestCase):
```
:+1: to tests

proteusvacuum committed:
use wrapped CouchUser and list comprehension

gcapalbo commented on corehq/apps/registration/utils.py:
```diff
@@ -92,27 +93,36 @@ def safe_unsubscribe_user_from_mailchimp_list(user, list_id, email=None):
-        list_id,
-        email=new_email,
-    )
+            users_subscribed_with_old_email[0] == user.get_id):
```
I think there's an edge case here that won't be caught.  Suppose two users, user A and user B, have the same email address.  User A is subscribed only to commcare users, and user B is subscribed only to the mass email list.  If user B changes their email, the email address will not be unsubscribed from the mass email list even though now there does not exist any user with that email address who is subscribed to the mass email list (same issue happens if user A changed email address, only the commcare users list would be wrong).  To fix it, I think the subscribed() check and subsequent unsubscribe would have to happen separately for each list.

proteusvacuum committed:
handle edge case

if users are subscribed to 'opposite' lists
